rfcs/2021-08-29-8381-vrl-iteration-support.md

```diff
@@ -90,28 +102,380 @@ One gap in the language right now is the possibility to _dynamically remap
+recursion is enabled, it will return the key of the to-be-recursed collection
+first, and then any items within that collection. Note that arrays are recursed
+as well, to allow recursing ”through” arrays into objects within those arrays.
+This allows for mapping _all_ keys in an object, even if those keys are deeply
+nested within objects within array(s) within the top-level object.
+
+##### `map_values`
+
+Map each individual value of an object or array to a different value.
+
+###### function signature
+
+```coffee
+map_values(value: object|array, recursive: bool) -> |any| { any }
+```
+
+###### details
+
+The function works similarly to `map_keys`, except that it maps the values
+instead of keys, and thus can also be used to map values within arrays.
+
+Recursion behaves similarly to `map_keys` as well.
+
+##### `for_each`
+
+Iterate over objects or arrays, without mutating any data.
+
+###### function signature
+
+```coffee
+for_each(value: object|array, recursive: bool) -> |string OR integer, any| { any }
+```
+
+###### details
+
+This can be considered a ”trap door” iteration function that allows you to
+tackle any use-case not solved by any of the existing (or future) specialized
+iteration functions.
+
+The drawback of such a function is that it potentially requires more manual
+”set-up” code to get the end-result (e.g. initializing empty collections to
+populate during a `for_each` run, for example).
+
+As the name implies, this function does not mutate the given collection, and
+instead always returns `null`. It can be used to mutate data external to the
+closure, while iterating over the collection. In a sense, it’s the most
+general-purpose iteration function that allows you to manually write mapping,
+reducing, filtering or counting logic.
+
+##### why no `map` function exists
+
+Some might note that there’s no `map` function, only specialized `map_keys` and
+`map_values`.
+
+The reason for this omission is that `map` becomes complicated when dealing
+with recursion, and the closure signature differs when dealing with an object or
+array, requiring us to know at compile-time the exact type of the iteration
+target.
+
+In the end, all current requested use-cases by operators could be solved by one
+of the three proposed iteration functions, allowing us to skip the additional
+work of figuring out how `map` would work exactly, until there’s an actual need
+for such a function (if ever).
+
+### Use Cases
+
+What follows is a list of reported use-cases, and a valid program that uses
+iteration to solve that use-case. Note that there are multiple ways to solve
+individual use-cases, this list shows one available solution per use-case.
+
+1. [nullify empty strings](https://discord.com/channels/742820443487993987/746070591097798688/950750550583050321)
+
+   ```coffee
+   . = map_values(., recursive: true) -> |value| { if value == "" { null } else { value } }
+   ```
```

Comment:
While I agree this is a good test case for iteration support, we should have functions for things like this (i.e., `compact`).

Reply:
I believe today's `compact` does this recursively today 🤔

Reply:
In this case, they don't want to delete the field, just map it to `null`.

Reply:
Isn't a deleted field equivalent to `null`? How would that be different in practice or for a downstream sink?

Reply:
I think the semantics of that vary by sink. `null` could be treated differently from undefined.

Reply:
I don't think that's the case today, but I digress, and this is unrelated to the design itself.